File: 1971-find-if-path-exists-in-graph/1971-find-if-path-exists-in-graph.py
class Solution:
    def validPath(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
        mp=defaultdict(list)
        for u,v in edges:
            mp[u].append(v)
            mp[v].append(u)
        q=collections.deque([source])
        vst=set()
        while q:
            cur=q.popleft()
            if cur==destination:
                return True
            
            for n in mp[cur]:
                if n in vst:
                    continue
                vst.add(n)
                q.append(n)
        return False
                    
            

File: 1971-find-if-path-exists-in-graph/1971-find-if-path-exists-in-graph.java
class Solution {
   public boolean validPath(int n, int[][] edges, int source, int destination) {
          HashMap<Integer, Stack<Integer>> graph = new HashMap<>();
        boolean[] visited = new boolean[n];
        for (int[] edge : edges) {
            graph.computeIfAbsent(edge[0], k -> new Stack<>()).push(edge[1]);
            graph.computeIfAbsent(edge[1], k -> new Stack<>()).push(edge[0]);
        }
        return dfs(graph, visited, source, destination);

    }

    private boolean dfs(HashMap<Integer, Stack<Integer>> graph, boolean[] visited, int source, int destination) {
        if (source == destination) {
            return true;
        }
        visited[source] = true;
        for (Integer next : graph.getOrDefault(source, new Stack<>())) {
            if (!visited[next]) {
                if (dfs(graph, visited, next, destination)) {
                    return true;
                }
            }
        }
        return false;
    }
}